Regression: Variant Ambiguity error during Gradle dependency resolution when apply the plugin lazily

## What version of OpenRewrite are you using?
I am using
- OpenRewrite Gradle plugin v6.28.1

## How are you running OpenRewrite?
I am using the Gradle plugin, and my project is a multi module project.

See a reproducer here: https://github.com/stefanroeck/openrewrite_reproducer
- checkout the reproducer repo
- run `./gradlew build`

In this project, two submodules exist where the one depends on the other. The open-rewrite plugin is applied lazily (in the original setup, we have hundreds of submodules and to this for maintenance reasons for many of such plugins). 

I believe this has been introduced with https://github.com/openrewrite/rewrite-gradle-plugin/commit/147c7effda5b9ef521ca6d8a341fde69d9881c56

## What did you expect to see?
- Build just runs fine

## What did you see instead?
Error message
```bash
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':SubModule2:test'.
> Could not resolve all dependencies for configuration ':SubModule2:testRuntimeClasspath'.
   > Could not resolve project :SubModule1.
     Required by:
         project :SubModule2
      > The consumer was configured to find a library for use during runtime, compatible with Java 17, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally. However we cannot choose between the following variants of project :SubModule1:
          - rewrite
          - runtimeElements
        All of them match the consumer attributes:
          - Variant 'rewrite' capability 'openRewriteReproducer:SubModule1:unspecified' declares a library for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally:
              - Unmatched attribute:
                  - Doesn't say anything about its target Java version (required compatibility with Java 17)
          - Variant 'runtimeElements' capability 'openRewriteReproducer:SubModule1:unspecified' declares a library for use during runtime, compatible with Java 17, packaged as a jar, and its dependencies declared externally:
              - Unmatched attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)

* Try:
> Ambiguity errors are explained in more detail at https://docs.gradle.org/8.11.1/userguide/variant_model.html#sub:variant-ambiguity.
> Review the variant matching algorithm at https://docs.gradle.org/8.11.1/userguide/variant_attributes.html#sec:abm_algorithm.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

```

--- a/plugin/src/main/java/org/openrewrite/gradle/RewritePlugin.java
+++ b/plugin/src/main/java/org/openrewrite/gradle/RewritePlugin.java
@@ -37,9 +37,7 @@ import org.jspecify.annotations.Nullable;
 import java.io.File;
 import java.util.Comparator;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Set;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static org.gradle.api.attributes.Bundling.BUNDLING_ATTRIBUTE;
@@ -72,43 +70,8 @@ public class RewritePlugin implements Plugin<Project> {
 
         // Rewrite module dependencies put here will be available to all rewrite tasks
         Configuration rewriteConf = project.getConfigurations().maybeCreate("rewrite");
-        // Defer actually evaluating the dependencies until resolution is triggered. This should allow
-        // the user to set the rewrite version in the extension. `addLater` doesn't work here because
-        // it operates on a single dependency at a time.
-        rewriteConf.getIncoming().beforeResolve(conf -> {
-            rewriteConf.getDependencies().addAll(
-                knownRewriteDependencies(extension, project.getDependencies())
-            );
-        });
 
-        // Because of how this Gradle has no criteria with which to select between variants of
-        // dependencies which expose differing capabilities. So those must be manually configured
-        try {
-            final ObjectFactory objectFactory = project.getObjects();
-            rewriteConf.attributes(attributes -> {
-                // Adapted from org.gradle.api.plugins.jvm.internal.DefaultJvmEcosystemAttributesDetails
-                attributes.attribute(
-                        Category.CATEGORY_ATTRIBUTE, objectFactory.named(Category.class, Category.LIBRARY));
-                attributes.attribute(Usage.USAGE_ATTRIBUTE, objectFactory.named(Usage.class, Usage.JAVA_RUNTIME));
-                attributes.attribute(
-                        LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE,
-                        objectFactory.named(LibraryElements.class, LibraryElements.JAR));
-                attributes.attribute(BUNDLING_ATTRIBUTE, objectFactory.named(Bundling.class, Bundling.EXTERNAL));
-                try {
-                    attributes.attribute(
-                            TARGET_JVM_ENVIRONMENT_ATTRIBUTE,
-                            objectFactory.named(TargetJvmEnvironment.class, TargetJvmEnvironment.STANDARD_JVM));
-                } catch (final NoClassDefFoundError ex) {
-                    // Old versions of Gradle don't have the class TargetJvmEnvironment and that's OK, we can always
-                    // try this attribute instead
-                    attributes.attribute(Attribute.of("org.gradle.jvm.environment", String.class), "standard-jvm");
-                }
-            });
-        } catch (final NoClassDefFoundError ex) {
-            // Old versions of Gradle don't have all of these attributes and that's OK
-        }
-
-        Provider<Set<File>> resolvedDependenciesProvider = project.provider(() -> getResolvedDependencies(rewriteConf));
+        Provider<Set<File>> resolvedDependenciesProvider = project.provider(() -> getResolvedDependencies(project, extension, rewriteConf));
 
         TaskProvider<RewriteRunTask> rewriteRun = project.getTasks().register("rewriteRun", RewriteRunTask.class, task -> {
             task.setExtension(extension);
@@ -190,16 +153,43 @@ public class RewritePlugin implements Plugin<Project> {
         });
     }
 
-    private Set<File> getResolvedDependencies(Configuration rewriteConf) {
-        if (resolvedDependencies != null) {
-            return resolvedDependencies;
-        }
+    private Set<File> getResolvedDependencies(Project project, RewriteExtension extension, Configuration rewriteConf) {
+        if (resolvedDependencies == null) {
+            Dependency[] dependencies = Stream.concat(
+                    knownRewriteDependencies(extension, project.getDependencies()),
+                    rewriteConf.getDependencies().stream()
+            ).toArray(Dependency[]::new);
+            // By using a detached configuration, we separate this dependency resolution from the rest of the project's
+            // configuration. This also means that Gradle has no criteria with which to select between variants of
+            // dependencies which expose differing capabilities. So those must be manually configured
+            Configuration detachedConf = project.getConfigurations().detachedConfiguration(dependencies);
 
-        resolvedDependencies = rewriteConf.resolve();
+            try {
+                ObjectFactory objectFactory = project.getObjects();
+                detachedConf.attributes(attributes -> {
+                    // Adapted from org.gradle.api.plugins.jvm.internal.DefaultJvmEcosystemAttributesDetails
+                    attributes.attribute(Category.CATEGORY_ATTRIBUTE, objectFactory.named(Category.class, Category.LIBRARY));
+                    attributes.attribute(Usage.USAGE_ATTRIBUTE, objectFactory.named(Usage.class, Usage.JAVA_RUNTIME));
+                    attributes.attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objectFactory.named(LibraryElements.class, LibraryElements.JAR));
+                    attributes.attribute(BUNDLING_ATTRIBUTE, objectFactory.named(Bundling.class, Bundling.EXTERNAL));
+                    try {
+                        attributes.attribute(TARGET_JVM_ENVIRONMENT_ATTRIBUTE, objectFactory.named(TargetJvmEnvironment.class, TargetJvmEnvironment.STANDARD_JVM));
+                    } catch (NoClassDefFoundError e) {
+                        // Old versions of Gradle don't have the class TargetJvmEnvironment and that's OK, we can always
+                        // try this attribute instead
+                        attributes.attribute(Attribute.of("org.gradle.jvm.environment", String.class), "standard-jvm");
+                    }
+                });
+            } catch (NoClassDefFoundError e) {
+                // Old versions of Gradle don't have all of these attributes and that's OK
+            }
+
+            resolvedDependencies = detachedConf.resolve();
+        }
         return resolvedDependencies;
     }
 
-    private static List<Dependency> knownRewriteDependencies(RewriteExtension extension, DependencyHandler deps) {
+    private static Stream<Dependency> knownRewriteDependencies(RewriteExtension extension, DependencyHandler deps) {
         String rewriteVersion = extension.getRewriteVersion();
         return Stream.of(
                 deps.create("org.openrewrite:rewrite-core:" + rewriteVersion),
@@ -222,6 +212,6 @@ public class RewritePlugin implements Plugin<Project> {
                 deps.create("org.openrewrite.gradle.tooling:model:" + extension.getRewriteGradleModelVersion()),
                 deps.create("com.fasterxml.jackson.module:jackson-module-kotlin:" + extension.getJacksonModuleKotlinVersion()),
                 deps.create("com.fasterxml.jackson.datatype:jackson-datatype-jsr310:" + extension.getJacksonModuleKotlinVersion())
-        ).collect(Collectors.toList());
+        );
     }
 }

--- a/plugin/src/test/kotlin/org/openrewrite/gradle/RewritePluginTest.kt
+++ b/plugin/src/test/kotlin/org/openrewrite/gradle/RewritePluginTest.kt
@@ -72,7 +72,8 @@ interface RewritePluginTest {
         @TempDir projectDir: File
     ) {
         gradleProject(projectDir) {
-            buildGradle("""
+            buildGradle(
+                """
                 plugins {
                     id("org.openrewrite.rewrite")
                 }
@@ -84,7 +85,8 @@ interface RewritePluginTest {
                        url = uri("https://oss.sonatype.org/content/repositories/snapshots")
                     }
                 }
-            """)
+                """
+            )
         }
         val result = runGradle(projectDir, taskName(), "--configuration-cache")
         val taskResult = result.task(":${taskName()}")!!

--- a/plugin/src/test/kotlin/org/openrewrite/gradle/RewriteRunTest.kt
+++ b/plugin/src/test/kotlin/org/openrewrite/gradle/RewriteRunTest.kt
@@ -1595,7 +1595,7 @@ class RewriteRunTest : RewritePluginTest {
                 }
 
                 apply from: 'dependencies.gradle'
-            """
+                """
             )
             otherGradleScript(
                 "dependencies.gradle", """
@@ -1609,7 +1609,7 @@ class RewriteRunTest : RewritePluginTest {
                 dependencies {
                     implementation("com.fasterxml.jackson.core:jackson-databind:2.15.2")
                 }
-            """
+                """
             )
         }
         projectDir.resolve("build").mkdirs()

